src/main/java/com/squareup/javawriter/builders/CodeWriter.java

```diff
@@ -101,15 +132,47 @@ public CodeWriter emit(Snippet snippet) {
     return this;
   }
 
+  private void emitLiteral(Object o) {
+    if (o instanceof TypeSpec) {
+      TypeSpec typeSpec = (TypeSpec) o;
+      typeSpec.emit(this);
+    } else {
+      emitAndIndent(String.valueOf(o));
+    }
+  }
+
   private void emitType(Object arg) {
     TypeName typeName = toTypeName(arg);
     emittedTypes.add(typeName);
 
-    String shortName = !visibleTypes.contains(typeName)
-        ? importedTypes.get(typeName)
-        : null;
-
-    emitAndIndent(shortName != null ? shortName : typeName.toString());
+    // TODO(jwilson): replace instanceof nonsense with polymorphism!
+    if (typeName instanceof ParameterizedTypeName) {
+      ParameterizedTypeName parameterizedTypeName = (ParameterizedTypeName) typeName;
+      emitType(parameterizedTypeName.type());
+      emitAndIndent("<");
+      boolean firstParameter = true;
+      for (TypeName parameter : parameterizedTypeName.parameters()) {
+        if (!firstParameter) emitAndIndent(", ");
+        emitType(parameter);
+        firstParameter = false;
+      }
+      emitAndIndent(">");
+    } else if (typeName instanceof WildcardName) {
+      WildcardName wildcardName = (WildcardName) typeName;
+      TypeName extendsBound = wildcardName.extendsBound();
+      TypeName superBound = wildcardName.superBound();
+      if (extendsBound != null) {
+        emit("? extends $T", extendsBound);
+      } else if (superBound != null) {
+        emit("? super $T", superBound);
+      }
+      // TODO(jwilson): special case ? for List<?>.
```

Comment:
This is just an `else` on the previous `if`. Both will be null on a boundless wildcard.